solidity/contracts/cryptography/AltBn128.sol

```diff
@@ -306,23 +318,39 @@ library AltBn128 {
      * the sum of two points on G1. Revert if the provided points aren't on the
      * curve.
      */
-    function g1Add(G1Point memory a, G1Point memory b) internal view returns (G1Point memory) {
-        uint256[4] memory arg;
-        arg[0] = a.x;
-        arg[1] = a.y;
-        arg[2] = b.x;
-        arg[3] = b.y;
-        uint256[2] memory c;
-
+    function g1Add(G1Point memory a, G1Point memory b)
+        internal view returns (G1Point memory c) {
         /* solium-disable-next-line */
         assembly {
+            let arg := mload(0x40)
+            mstore(arg, mload(a))
+            mstore(add(arg, 0x20), mload(add(a, 0x20)))
+            mstore(add(arg, 0x40), mload(b))
+            mstore(add(arg, 0x60), mload(add(b, 0x20)))
             // 0x60 is the ECADD precompile address
             if iszero(staticcall(not(0), 0x06, arg, 0x80, c, 0x40)) {
                 revert(0, 0)
             }
         }
+    }
 
-        return G1Point(c[0], c[1]);
+    function _g1Add(uint256 ax, uint256 ay, uint256 bx, uint256 by)
```

Comment:
Do we need this function? I think it's not used anywhere.

Reply:
Neither `g1Add` nor `_g1Add` is used in the library, but I made both the memory-allocating and the stack-only versions available in the API. This makes the API a bit crowded but I'm not sure what can be done to improve it, except maybe making the stack-only functions `private` and only used for optimization inside the library.